File: services/stats-svc/momentslite.go
package main

import (
	"fmt"
	"net/http"

	"github.com/google/uuid"
)

// Moments-lite indeksi: statik sitenin harita-başına kompakt olay dizini.
// Exporter bu ucu her harita için sayfalar; istemci (lib/momentslite.ts)
// aynı DSL alt kümesini (kill/grenade/bomb/economy) bu sütunlu veri
// üzerinde değerlendirir. presence CH gerektirdiğinden stüdyoda kalır.
// Sütun sözlükleri (buy/side/grenade sırası) istemciyle BİREBİR sözleşmedir.

type miMatch struct {
	ID  uuid.UUID `json:"id"`
	Src string    `json:"src"`
}

type miRounds struct {
	M  []int32   `json:"m"`  // matches dizinine indeks
	N  []int16   `json:"n"`  // round_number
	TB []int8    `json:"tb"` // buy: 0=yok, 1..5 = pistol,eco,semi,force,full
	CB []int8    `json:"cb"`
	TE []int32   `json:"te"` // equip value (-1 = bilinmiyor; SQL NULL karşılığı)
	CE []int32   `json:"ce"`
	FE []int32   `json:"fe"` // freeze_end_tick
	BS []int8    `json:"bs"` // bomb site: 0=yok, 1=A, 2=B
	BP []int32   `json:"bp"` // bomb_plant_tick (0 = plant yok)
	ER []int16   `json:"er"` // end_reasons dizinine indeks
	ET []int32   `json:"et"` // end_tick
}

type miKills struct {
	RI []int32   `json:"ri"` // rounds satır indeksi
	T  []int32   `json:"t"`
	RT []float32 `json:"rt"`
	W  []int32   `json:"w"`  // weapons dizinine indeks
	F  []int8    `json:"f"`  // bit: 1=first_kill, 2=trade, 4=headshot
	AP []int32   `json:"ap"` // places: 0=yok, i+1
	VP []int32   `json:"vp"`
	A  []int32   `json:"a"`  // players: 0=yok, i+1 (saldırgan)
	S  []int8    `json:"s"`  // 0=bilinmiyor, 1=T, 2=CT (saldırgan tarafı)
}

type miGrenades struct {
	RI []int32 `json:"ri"`
	T  []int32 `json:"t"` // detonate_tick
	G  []int8  `json:"g"` // 1..6 = flash,smoke,he,molotov,incendiary,decoy
	S  []int8  `json:"s"`
	F  []int8  `json:"f"` // bit: 1=first_of_type_in_round
	P  []int32 `json:"p"` // players: 0=yok, i+1 (atan)
}

type momentsIndex struct {
	V          int        `json:"v"`
	Map        string     `json:"map"`
	Matches    []miMatch  `json:"matches"`
	Weapons    []string   `json:"weapons"`
	Places     []string   `json:"places"`
	Players    []string   `json:"players"`
	EndReasons []string   `json:"end_reasons"`
	Rounds     miRounds   `json:"rounds"`
	Kills      miKills    `json:"kills"`
	Grenades   miGrenades `json:"grenades"`
}

// interner: string → 1-tabanlı sözlük indeksi ("" → 0)
type interner struct {
	idx  map[string]int32
	list []string
}

func newInterner() *interner { return &interner{idx: map[string]int32{}} }

func (in *interner) get(s string) int32 {
	if s == "" {
		return 0
	}
	if i, ok := in.idx[s]; ok {
		return i
	}
	in.list = append(in.list, s)
	in.idx[s] = int32(len(in.list))
	return in.idx[s]
}

var miBuy = map[string]int8{"pistol": 1, "eco": 2, "semi": 3, "force": 4, "full": 5}
var miGren = map[string]int8{"flash": 1, "smoke": 2, "he": 3, "molotov": 4, "incendiary": 5, "decoy": 6}

func miSide(s string) int8 {
	switch s {
	case "T":
		return 1
	case "CT":
		return 2
	}
	return 0
}

// GET /api/v1/export/moments-index?map=de_mirage
func (s *server) momentsIndexHandler(w http.ResponseWriter, r *http.Request) {
	mapName := r.URL.Query().Get("map")
	if mapName == "" {
		writeErr(w, 400, fmt.Errorf("map is required"))
		return
	}
	ctx := r.Context()
	out := &momentsIndex{V: 1, Map: mapName}
	weapons := newInterner()
	places := newInterner()
	players := newInterner()
	endReasons := newInterner()
	matchIdx := map[uuid.UUID]int32{}
	roundIdx := map[[2]int64]int32{} // [matchIdx, roundNumber] → rounds satırı

	rrows, err := s.pg.Query(ctx, `
		SELECT r.match_id, COALESCE(m.source,''), r.round_number,
		       COALESCE(r.freeze_end_tick,0), COALESCE(r.end_tick,0),
		       COALESCE(r.end_reason,''), COALESCE(r.bomb_plant_tick,0),
		       COALESCE(r.bomb_site,''), COALESCE(r.t_equip_value,-1),
		       COALESCE(r.ct_equip_value,-1), COALESCE(r.t_buy_type,''),
		       COALESCE(r.ct_buy_type,'')
		FROM rounds r JOIN matches m ON m.match_id = r.match_id
		WHERE m.status = 'ready' AND m.map_name = $1
		ORDER BY r.match_id, r.round_number`, mapName)
	if err != nil {
		writeErr(w, 500, err)
		return
	}
	defer rrows.Close()
	for rrows.Next() {
		var mid uuid.UUID
		var src, er, bs, tb, cb string
		var n int16
		var fe, et, bp, te, ce int32
		if err := rrows.Scan(&mid, &src, &n, &fe, &et, &er, &bp, &bs, &te, &ce, &tb, &cb); err != nil {
			writeErr(w, 500, err)
			return
		}
		mi, ok := matchIdx[mid]
		if !ok {
			mi = int32(len(out.Matches))
			matchIdx[mid] = mi
			out.Matches = append(out.Matches, miMatch{ID: mid, Src: src})
		}
		roundIdx[[2]int64{int64(mi), int64(n)}] = int32(len(out.Rounds.M))
		var bsi int8
		switch bs {
		case "A":
			bsi = 1
		case "B":
			bsi = 2
		}
		ro := &out.Rounds
		ro.M = append(ro.M, mi)
		ro.N = append(ro.N, n)
		ro.TB = append(ro.TB, miBuy[tb])
		ro.CB = append(ro.CB, miBuy[cb])
		ro.TE = append(ro.TE, te)
		ro.CE = append(ro.CE, ce)
		ro.FE = append(ro.FE, fe)
		ro.BS = append(ro.BS, bsi)
		ro.BP = append(ro.BP, bp)
		ro.ER = append(ro.ER, int16(endReasons.get(er)))
		ro.ET = append(ro.ET, et)
	}
	if err := rrows.Err(); err != nil {
		writeErr(w, 500, err)
		return
	}

	krows, err := s.pg.Query(ctx, `
		SELECT k.match_id, k.round_number, COALESCE(k.tick,0),
		       COALESCE(k.round_time,0), COALESCE(lower(k.weapon),''),
		       COALESCE(k.is_first_kill,false), COALESCE(k.is_trade,false),
		       COALESCE(k.headshot,false), COALESCE(k.attacker_place,''),
		       COALESCE(k.victim_place,''), COALESCE(p.nickname,''),
		       COALESCE(s.side,'')
		FROM kills k
		JOIN matches m ON m.match_id = k.match_id
		     AND m.status = 'ready' AND m.map_name = $1
		LEFT JOIN players p ON p.player_id = k.attacker_id
		LEFT JOIN player_round_states s ON s.match_id = k.match_id
		     AND s.round_number = k.round_number AND s.player_id = k.attacker_id
		ORDER BY k.match_id, k.round_number, k.tick`, mapName)
	if err != nil {
		writeErr(w, 500, err)
		return
	}
	defer krows.Close()
	for krows.Next() {
		var mid uuid.UUID
		var n int16
		var tick int32
		var rt float32
		var weapon, ap, vp, nick, side string
		var first, trade, hs bool
		if err := krows.Scan(&mid, &n, &tick, &rt, &weapon, &first, &trade, &hs, &ap, &vp, &nick, &side); err != nil {
			writeErr(w, 500, err)
			return
		}
		mi, ok := matchIdx[mid]
		if !ok {
			continue
		}
		ri, ok := roundIdx[[2]int64{int64(mi), int64(n)}]
		if !ok {
			continue
		}
		var f int8
		if first {
			f |= 1
		}
		if trade {
			f |= 2
		}
		if hs {
			f |= 4
		}
		k := &out.Kills
		k.RI = append(k.RI, ri)
		k.T = append(k.T, tick)
		k.RT = append(k.RT, rt)
		k.W = append(k.W, weapons.get(weapon))
		k.F = append(k.F, f)
		k.AP = append(k.AP, places.get(ap))
		k.VP = append(k.VP, places.get(vp))
		k.A = append(k.A, players.get(nick))
		k.S = append(k.S, miSide(side))
	}
	if err := krows.Err(); err != nil {
		writeErr(w, 500, err)
		return
	}

	grows, err := s.pg.Query(ctx, `
		SELECT g.match_id, g.round_number, g.detonate_tick, g.type,
		       COALESCE(g.side,''), COALESCE(g.is_first_of_type_in_round,false),
		       COALESCE(p.nickname,'')
		FROM grenades g
		JOIN matches m ON m.match_id = g.match_id
		     AND m.status = 'ready' AND m.map_name = $1
		LEFT JOIN players p ON p.player_id = g.thrower_id
		WHERE g.detonate_tick IS NOT NULL
		ORDER BY g.match_id, g.round_number, g.detonate_tick`, mapName)
	if err != nil {
		writeErr(w, 500, err)
		return
	}
	defer grows.Close()
	for grows.Next() {
		var mid uuid.UUID
		var n int16
		var tick int32
		var typ, side, nick string
		var first bool
		if err := grows.Scan(&mid, &n, &tick, &typ, &side, &first, &nick); err != nil {
			writeErr(w, 500, err)
			return
		}
		mi, ok := matchIdx[mid]
		if !ok {
			continue
		}
		ri, ok := roundIdx[[2]int64{int64(mi), int64(n)}]
		if !ok {
			continue
		}
		var f int8
		if first {
			f |= 1
		}
		g := &out.Grenades
		g.RI = append(g.RI, ri)
		g.T = append(g.T, tick)
		g.G = append(g.G, miGren[typ])
		g.S = append(g.S, miSide(side))
		g.F = append(g.F, f)
		g.P = append(g.P, players.get(nick))
	}
	if err := grows.Err(); err != nil {
		writeErr(w, 500, err)
		return
	}

	out.Weapons = weapons.list
	out.Places = places.list
	out.Players = players.list
	out.EndReasons = endReasons.list
	writeJSON(w, 200, out)
}

File: apps/web/src/lib/momentslite.ts
// Moments-lite: statik sitede DSL alt kümesi. Sunucudaki dsl/compile.go'nun
// kill/grenade/bomb/economy dallarını, exporter'ın harita-başına yayınladığı
// kompakt indeks (services/stats-svc/momentslite.go) üzerinde koşar.
// presence CH tick verisi ister — stüdyoda kalır. Sözlük sıraları (buy,
// side, grenade tipi) sunucu tarafıyla BİREBİR sözleşmedir.
import type { QueryResult, Clip } from '../api';
import { staticGet } from './staticdata';

const TICK_RATE = 64;
const BUYS = ['', 'pistol', 'eco', 'semi', 'force', 'full'];
const GRENS = ['', 'flash', 'smoke', 'he', 'molotov', 'incendiary', 'decoy'];
const SIDES = ['', 'T', 'CT'];

interface MomentsIndex {
  v: number;
  map: string;
  matches: { id: string; src: string }[];
  weapons: string[];
  places: string[];
  players: string[];
  end_reasons: string[];
  rounds: {
    m: number[]; n: number[]; tb: number[]; cb: number[]; te: number[];
    ce: number[]; fe: number[]; bs: number[]; bp: number[]; er: number[]; et: number[];
  };
  kills: {
    ri: number[]; t: number[]; rt: number[]; w: number[]; f: number[];
    ap: number[]; vp: number[]; a: number[]; s: number[];
  };
  grenades: { ri: number[]; t: number[]; g: number[]; s: number[]; f: number[]; p: number[] };
}

// DSL Query (dsl.go ile aynı şekil; yalnız kullanılan alanlar)
export interface LiteQuery {
  intent: string;
  filters: {
    map?: string;
    side?: string;
    buy_type?: string[];
    round_number?: { min?: number; max?: number };
    source?: string;
    player?: { nickname?: string };
    event?: {
      type: string;
      weapon?: string; first_kill?: boolean; trade?: boolean; headshot?: boolean;
      area?: string; area_of?: string;
      grenade_type?: string; order?: string;
      bomb_action?: string; site?: string;
      equip_min?: number; equip_max?: number;
      time_window?: { from: number; to: number };
    };
  };
  output?: { format?: string; context_seconds?: number[]; metric?: string };
}

const cache = new Map<string, Promise<MomentsIndex>>();
function getIndex(map: string): Promise<MomentsIndex> {
  let p = cache.get(map);
  if (!p) {
    p = staticGet<MomentsIndex>(`/api/v1/export/moments-index?map=${encodeURIComponent(map)}`);
    p.catch(() => cache.delete(map));
    cache.set(map, p);
  }
  return p;
}

interface Moment { ri: number; tick: number; rt: number }

// roundFilter karşılığı: filtreyi geçen rounds satır indeks kümesi
function roundSet(idx: MomentsIndex, f: LiteQuery['filters']): Set<number> {
  const r = idx.rounds;
  const out = new Set<number>();
  const buyIdx = (f.buy_type ?? []).map((b) => BUYS.indexOf(b));
  for (let i = 0; i < r.m.length; i++) {
    if (f.source && idx.matches[r.m[i]].src !== f.source) continue;
    const rn = f.round_number;
    if (rn?.min && r.n[i] < rn.min) continue;
    if (rn?.max && r.n[i] > rn.max) continue;
    if (buyIdx.length) {
      const ok =
        f.side === 'T' ? buyIdx.includes(r.tb[i]) :
        f.side === 'CT' ? buyIdx.includes(r.cb[i]) :
        buyIdx.includes(r.tb[i]) || buyIdx.includes(r.cb[i]);
      if (!ok) continue;
    }
    out.add(i);
  }
  return out;
}

function momentsFor(idx: MomentsIndex, q: LiteQuery, rounds: Set<number>): Moment[] {
  const ev = q.filters.event!;
  const f = q.filters;
  const out: Moment[] = [];
  const r = idx.rounds;

  if (ev.type === 'kill') {
    const k = idx.kills;
    const wantSide = SIDES.indexOf(f.side ?? '') > 0 ? SIDES.indexOf(f.side!) : 0;
    const wantW = ev.weapon ? idx.weapons.indexOf(ev.weapon.toLowerCase()) + 1 : 0;
    if (ev.weapon && wantW === 0) return []; // silah arşivde hiç yok
    const wantArea = ev.area ? idx.places.indexOf(ev.area) + 1 : 0;
    if (ev.area && wantArea === 0) return [];
    const wantA = f.player?.nickname
      ? idx.players.findIndex((p) => p.toLowerCase() === f.player!.nickname!.toLowerCase()) + 1
      : 0;
    if (f.player?.nickname && wantA === 0) return [];
    for (let i = 0; i < k.ri.length; i++) {
      if (!rounds.has(k.ri[i])) continue;
      if (wantSide && k.s[i] !== wantSide) continue;
      if (wantW && k.w[i] !== wantW) continue;
      if (ev.first_kill != null && !!(k.f[i] & 1) !== ev.first_kill) continue;
      if (ev.trade != null && !!(k.f[i] & 2) !== ev.trade) continue;
      if (ev.headshot != null && !!(k.f[i] & 4) !== ev.headshot) continue;
      if (wantArea && (ev.area_of === 'attacker' ? k.ap[i] : k.vp[i]) !== wantArea) continue;
      const tw = ev.time_window;
      if (tw && (k.rt[i] < tw.from || k.rt[i] > tw.to)) continue;
      if (wantA && k.a[i] !== wantA) continue;
      out.push({ ri: k.ri[i], tick: k.t[i], rt: k.rt[i] });
    }
    return out;
  }

  if (ev.type === 'grenade') {
    const g = idx.grenades;
    const wantSide = SIDES.indexOf(f.side ?? '') > 0 ? SIDES.indexOf(f.side!) : 0;
    const wantG = ev.grenade_type ? GRENS.indexOf(ev.grenade_type) : 0;
    const wantP = f.player?.nickname
      ? idx.players.findIndex((p) => p.toLowerCase() === f.player!.nickname!.toLowerCase()) + 1
      : 0;
    if (f.player?.nickname && wantP === 0) return [];
    for (let i = 0; i < g.ri.length; i++) {
      if (!rounds.has(g.ri[i])) continue;
      if (wantSide && g.s[i] !== wantSide) continue;
      if (wantG && g.g[i] !== wantG) continue;
      if (ev.order === 'first_of_type_in_round' && !(g.f[i] & 1)) continue;
      const rt = (g.t[i] - r.fe[g.ri[i]]) / TICK_RATE;
      const tw = ev.time_window;
      if (tw && (rt < tw.from || rt > tw.to)) continue;
      if (wantP && g.p[i] !== wantP) continue;
      out.push({ ri: g.ri[i], tick: g.t[i], rt: Math.max(rt, 0) });
    }
    return out;
  }

  if (ev.type === 'bomb') {
    const wantSite = ev.site === 'A' ? 1 : ev.site === 'B' ? 2 : 0;
    const act = ev.bomb_action || 'plant';
    const wantER = act === 'defuse' ? idx.end_reasons.indexOf('bomb_defused') + 1
      : act === 'explode' ? idx.end_reasons.indexOf('bomb_exploded') + 1 : 0;
    if (act !== 'plant' && wantER === 0) return [];
    for (const i of rounds) {
      if (wantSite && r.bs[i] !== wantSite) continue;
      let tick: number;
      if (act === 'plant') {
        if (!r.bp[i]) continue;
        tick = r.bp[i];
      } else {
        if (r.er[i] !== wantER) continue;
        tick = r.et[i];
      }
      const rt = Math.max((tick - r.fe[i]) / TICK_RATE, 0);
      const tw = ev.time_window;
      if (tw && (rt < tw.from || rt > tw.to)) continue;
      out.push({ ri: i, tick, rt });
    }
    out.sort((a, b) => a.ri - b.ri);
    return out;
  }

  if (ev.type === 'economy') {
    for (const i of rounds) {
      const v = f.side === 'CT' ? r.ce[i] : r.te[i];
      if (v < 0) continue; // SQL NULL: sunucuda eşik karşılaştırması geçmez
      if (ev.equip_min != null && v < ev.equip_min) continue;
      if (ev.equip_max != null && v > ev.equip_max) continue;
      out.push({ ri: i, tick: r.fe[i], rt: 0 });
    }
    out.sort((a, b) => a.ri - b.ri);
    return out;
  }

  throw new Error(
    'presence queries need ClickHouse tick data — run the self-hosted studio for this one',
  );
}

export async function liteQuery(raw: unknown): Promise<QueryResult> {
  const q = raw as LiteQuery;
  const start = performance.now();
  const f = q.filters ?? {};
  if (!f.map) {
    throw new Error('pick a map first — on the static site the archive is indexed per map');
  }
  if (f.event?.type === 'presence') {
    throw new Error(
      'presence queries need ClickHouse tick data — run the self-hosted studio for this one',
    );
  }
  const idx = await getIndex(f.map);
  const rounds = roundSet(idx, f);
  const res: QueryResult = { intent: q.intent, duration_ms: 0 };

  if (q.intent === 'heatmap_filterset' || !f.event) {
    res.round_count = rounds.size;
    if (q.intent === 'aggregate') res.count = rounds.size;
    res.duration_ms = Math.round(performance.now() - start);
    return res;
  }

  const moments = momentsFor(idx, q, rounds);
  res.round_count = rounds.size;
  const fmt = q.output?.format ?? (q.intent === 'aggregate' ? 'aggregate' : 'clips');
  const r = idx.rounds;

  if (fmt === 'clips') {
    const cs = q.output?.context_seconds ?? [5, 8];
    const pre = Math.round(cs[0] * TICK_RATE);
    const post = Math.round(cs[1] * TICK_RATE);
    res.clips = moments.map((m): Clip => ({
      match_id: idx.matches[r.m[m.ri]].id,
      map_name: idx.map,
      round_number: r.n[m.ri],
      tick: m.tick,
      round_time: m.rt,
      tick_start: Math.max(0, m.tick - pre),
      tick_end: m.tick + post,
    }));
  } else if (fmt === 'rounds') {
    const seen = new Set<number>();
    res.rounds = [];
    for (const m of moments) {
      if (seen.has(m.ri)) continue;
      seen.add(m.ri);
      res.rounds.push({
        match_id: idx.matches[r.m[m.ri]].id,
        map_name: idx.map,
        round_number: r.n[m.ri],
      });
    }
  } else {
    res.count = moments.length;
    if ((q.output?.metric ?? 'count') === 'per_round' && rounds.size > 0) {
      res.per_round = moments.length / rounds.size;
    }
  }
  res.duration_ms = Math.round(performance.now() - start);
  return res;
}
